clues_test.go

```diff
@@ -424,6 +434,154 @@ func TestAddTraceNameTo(t *testing.T) {
+		expectM  msa
+		expectS  sa
+	}{
+		{
+			name:     "single",
+			kvs:      [][]string{{"k", "v"}},
+			isolated: false,
+			expectM:  msa{"k": "v"},
+			expectS:  sa{"k", "v"},
+		},
+		{
+			name:     "multiple",
+			kvs:      [][]string{{"a", "1"}, {"b", "2"}},
+			isolated: false,
+			expectM:  msa{"k": "v", "a": "1", "b": "2"},
+			expectS:  sa{"k", "v", "a", "1", "b", "2"},
+		},
+		{
+			name:     "duplicates",
+			kvs:      [][]string{{"a", "1"}, {"a", "2"}},
+			isolated: false,
+			expectM:  msa{"k": "v", "a": "2"},
+			expectS:  sa{"k", "v", "a", "2"},
+		},
+		{
+			name:     "single isolated",
+			kvs:      [][]string{{"k", "v"}},
+			isolated: true,
+			expectM:  msa{"k": "v", "proxy_pfx_k": "v"},
+			expectS:  sa{"k", "v", "proxy_pfx_k", "v"},
+		},
+		{
+			name:     "multiple isolated",
+			kvs:      [][]string{{"a", "1"}, {"b", "2"}},
+			isolated: true,
+			expectM:  msa{"k": "v", "proxy_pfx_a": "1", "proxy_pfx_b": "2"},
+			expectS:  sa{"k", "v", "proxy_pfx_a", "1", "proxy_pfx_b", "2"},
+		},
+		{
+			name:     "duplicates isolated",
+			kvs:      [][]string{{"a", "1"}, {"a", "2"}},
+			isolated: true,
+			expectM:  msa{"k": "v", "proxy_pfx_a": "2"},
+			expectS:  sa{"k", "v", "proxy_pfx_a", "2"},
+		},
+	}
+	for _, test := range table {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.WithValue(context.Background(), testCtx{}, "instance")
+			ctx = clues.Add(ctx, "k", "v")
+
+			check := msa{"k": "v"}
+
+			mustEquals(t, check, clues.In(ctx).Map(), true)
+
+			ctx = clues.AddProxy(ctx, "proxy_pfx", test.isolated)
+
+			for _, kv := range test.kvs {
+				clues.Add(ctx, kv[0], kv[1])
+
+				if test.isolated {
+					check["proxy_pfx_"+kv[0]] = kv[1]
+				} else {
+					check[kv[0]] = kv[1]
+				}
+
+				mustEquals(t, check, clues.In(ctx).Map(), true)
+			}
+
+			assert(
+				t, ctx, "",
+				test.expectM, msa{},
+				test.expectS, sa{})
+		})
+	}
+}
+
+func TestProxy_multipleLayers(t *testing.T) {
+	ctx := clues.Add(context.Background(), 1, 1)
+	pxyCtx := clues.AddProxy(ctx, "p1", true)
+
+	t.Run("1", func(t *testing.T) {
+		mapEquals(t, ctx, msa{"1": 1})
+		mapEquals(t, pxyCtx, msa{"1": 1})
+	})
+
+	ctx2 := clues.Add(pxyCtx, 2, 2)
+	pxyCtx2 := clues.AddProxy(ctx2, "p2", true)
+
+	t.Run("2", func(t *testing.T) {
+		mapEquals(t, ctx, msa{"1": 1})
+		mapEquals(t, pxyCtx, msa{
+			"1":    1,
+			"p1_2": 2,
+		})
+		mapEquals(t, ctx2, msa{
+			"1":    1,
+			"2":    2,
+			"p1_2": 2,
+		})
+		mapEquals(t, pxyCtx2, msa{
+			"1": 1,
+			"2": 2,
+		})
+	})
+
+	ctx3 := clues.Add(pxyCtx2, 3, 3)
+
+	t.Run("3", func(t *testing.T) {
+		mapEquals(t, ctx, msa{"1": 1})
+		// proxy 1 is still accumulating all the values added
+		// to descendant contexts, even after proxy 2 replaces
+		// it, because proxy 2 should maintain an ancestry ref
+		// to proxy 1.
+		mapEquals(t, pxyCtx, msa{
+			"1":    1,
+			"p1_2": 2,
+			"p1_3": 3,
+		})
+		mapEquals(t, ctx2, msa{
+			"1":    1,
+			"2":    2,
+			"p1_2": 2,
+			"p1_3": 3,
```

Comment:
I think this could turn into a point of confusion that makes it difficult to track how proxied values will appear. For instance, this value was added via proxy `p2`, but shows up with the prefix `p1`. So we know longer know that it came from `p2` nor is it apparent how it got to this context.

Reply:
Ah, I think the test variables might be making this confusing.  We're checking `ctx2`, which does not contain a reference to `proxy2`.  So when we _extract_ values from the context, it's all proxy1, because proxy2 doesn't exist in the extraction.

Reply:
sure, but the call chain to actually get data into this proxy is
ctx3 -> pxy2 -> ctx2 -> pxy

The only reason these values exist in pxy and ctx2 is because they passed through pxy2. However, there's no information saying that they passed through pxy2 (i.e. in the form of a prefix representing a namespace) because all prefixes are local to the current proxy in the node. Therefore, there's no way to track which/how many proxies a value may have gone through, making it difficult to determine where the value may have been added (i.e. the data could have been added in any of the functions in the transitive closure of functions called from the function reading data out of the proxy and there's no additional data in the proxy to help narrow that down)

Reply:
> making it difficult to determine where the value may have been added

I suppose that has never been a concern here.  The pitch is that "the proxy gathers all data downstream of its addition".  There is no condition preventing additions of data following the addition of another proxy (more downsides with that choice, relatively).  Caring about where, downstream, the proxy gained a value would be a bit counter to its purpose.

Now, if your mind is around what happens within larger scale proxy trees, and the possibility of sibling value clobbering... yeah, I can see that as a problem.  But it goes back to the (difficult to enforce) misuse of the design.

Reply:
fair points